File: core/src/main/java/xtdb/vector/IVectorIndirection.java
package xtdb.vector;

import java.util.Arrays;

public interface IVectorIndirection {
    int valueCount();

    int getIndex(int idx);

    record Selection(int[] idxs) implements IVectorIndirection {

        @Override
        public int valueCount() {
            return idxs.length;
        }

        @Override
        public int getIndex(int idx) {
            return idxs[idx];
        }

        @Override
        public String toString() {
            return "(Selection {idxs=%s})".formatted(Arrays.toString(idxs));
        }
    }

    record Slice(int startIdx, int len) implements IVectorIndirection {
        @Override
        public int valueCount() {
            return len;
        }

        @Override
        public int getIndex(int idx) {
            return startIdx + idx;
        }

    }
}

File: core/src/main/kotlin/xtdb/vector/IVectorIndirection.kt
package xtdb.vector

interface IVectorIndirection {
    fun valueCount(): Int

    fun getIndex(idx: Int): Int

    operator fun get(idx: Int): Int {
        return getIndex(idx)
    }

    @Suppress("DEPRECATION")
    @JvmRecord
    data class Selection(val idxs: IntArray) : IVectorIndirection {
        override fun valueCount(): Int {
            return idxs.size
        }

        override fun getIndex(idx: Int): Int {
            return idxs[idx]
        }

        override fun toString(): String {
            return "(Selection {idxs=%s})".formatted(idxs.contentToString())
        }
    }

    data class Slice(val startIdx: Int, val len: Int) : IVectorIndirection {
        override fun valueCount(): Int {
            return len
        }

        override fun getIndex(idx: Int): Int {
            return startIdx + idx
        }
    }
}
